compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -1033,6 +1040,70 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         true
     }
 
+    fn suggest_add_clone_to_arg(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut Diagnostic,
+        trait_pred: ty::PolyTraitPredicate<'tcx>,
+    ) -> bool {
+        let span = obligation.cause.span;
+        let body_id = obligation.cause.body_id;
+        let self_ty = self.resolve_vars_if_possible(trait_pred.self_ty());
+        let ty = self.tcx.erase_late_bound_regions(self_ty);
+        let owner = self.tcx.hir().get_parent_item(body_id);
+        if let ObligationCauseCode::FunctionArgumentObligation { arg_hir_id, .. } = obligation.cause.code() &&
+            let arg_node = self.tcx.hir().get(*arg_hir_id) &&
+            let Node::Expr(Expr { kind: hir::ExprKind::Path(_), ..}) = arg_node &&
+            let Some(generics) = self.tcx.hir().get_generics(owner.def_id) &&
+            let ty::Ref(_, inner_ty, hir::Mutability::Not) = ty.kind() &&
+            let ty::Param(param) = inner_ty.kind() &&
+            let Some(generic_param) =
+                    generics.params.iter().find(|p| p.name.ident().as_str() == param.name.as_str())
+        {
+            let clone_trait = self.tcx.require_lang_item(LangItem::Clone, None);
+            let has_clone = self
+                .type_implements_trait(clone_trait, [ty], obligation.param_env)
+                .must_apply_modulo_regions();
+
+            let trait_pred_and_suggested_ty =
+                trait_pred.map_bound(|trait_pred| (trait_pred, *inner_ty));
+            let new_obligation = self.mk_trait_obligation_with_new_self_ty(
+                    obligation.param_env,
+                    trait_pred_and_suggested_ty,
+            );
+
+            if  has_clone && self.predicate_may_hold(&new_obligation) {
+                let clone_bound = generics.bounds_for_param(generic_param.def_id)
+                                                    .flat_map(|bp| bp.bounds)
+                                                    .any(|bound| {
+                    if let hir::GenericBound::Trait( hir::PolyTraitRef { trait_ref, ..}, ..) = bound {
```

Comment:
Can you check if the param ty implements clone using `type_implements_trait`, rather than searching through the bounds? We can have indirect `Clone` bounds like `T: Trait` where `trait Trait: Clone {}`

Reply:
Here I'm trying to check whether there is already `Clone` in bounds, to avoid suggest to add `Clone` in this scenario, https://github.com/rust-lang/rust/pull/106497/files#diff-9fae219c10889c5437dc06b57eaf0eed42613d7d7266debae6cc62ca9eb87e74R13-R15

```rust
fn bar_with_clone<T: X + Clone>(s: &T)  {
    foo(s);
}
```

Suggest to this is bad:

```rust
fn bar_with_clone<T: X + Clone + Clone>(s: &T)  {
    foo(s); 
}
```

Reply:
Yes, but you can check `type_implements_trait` with the param ty itself.

Reply:
Nice, fixed!